source/memory-store.ts

```diff
@@ -30,16 +20,25 @@ export default class MemoryStore implements Store {
 	windowMs!: number
 
 	/**
-	 * The map that stores the number of hits for each client in memory.
+	 * These two maps store usage (requests) and reset time by key (IP)
+	 *
+	 * They are split into two to avoid having to iterate through the entire set to determine which ones need reset.
+	 * Instead, Clients are moved from previous to current as new requests come in from them.
+	 * Once windowMS has elapsed, all clients left in previous are reset at once.
+	 *
 	 */
-	hits!: {
-		[key: string]: number | undefined
-	}
+	previous!: Map<string, Client>
+	current!: Map<string, Client>
 
 	/**
-	 * The time at which all hit counts will be reset.
+	 * Pool of unused clients, kept to reduce the number of objects created and destroyed
 	 */
-	resetTime!: Date
+	pool!: Client[]
+
+	/**
+	 * Maximum number of unused clients to keep in the pool
+	 */
+	poolSize = 100
```

Comment:
Maybe it should take [`os.freemem()`](https://nodejs.org/api/os.html#osfreemem) into account also, since it's essentially trading RAM usage for performance.

Reply:
Actually, inspecting with chrome dev tools shows that a single client object is only ~152 bytes, so 60k would be less than 1mb. So `os.freemem()` won't really have much affect.

I put together a quick test https://gist.github.com/nfriedly/fdba25bcf9f23d18fe8ac5bd1bad5dac and 5000 seems to be the winner from a few runs, but it honestly doesn't make much difference either way.